rfc#0001

```diff
@@ -0,0 +1,140 @@
+# WASM binary loading into an Enarx Keep
+- Authors: [Mike Bursell](https://github.com/MikeCamel)
+- Status: [PROPOSED](/README.md#proposed)
+- Since: 2020-01-10
+- Status Note: (has been subject to initial discussion and consensus)  
+- Supersedes: n/a
+- Start Date: 2020-02-16
+- Tags: architecture, webassembly
+
+## Summary
+
+This RFC describes the architecture adopted by Enarx for loading WASM
+packages into a prepared Keep. 
+
+## Motivation
+
+Once a Keep is prepared with microkernel, WASM runtime, etc., the
+workload (WASM binary + data) needs to be loaded into it.  A
+description is required.
+
+## Tutorial
+
+A Keep is the fundamental runtime component of Enarx: a TEE instance
++ Enarx runtime pieces, including WebAssembly and WASI layers.  In
+order for it to execute a workload, the WASM (WebAssembly) package
+that comprises it (binary/binaries + data elements) need to be loaded
+into it.  This is sent to it by the Enarx client agent (ECA),
+encrypted under a session key established at Keep provisioning time,
+and communicated to the ECA - the workload provisioning encryption
+key (WPEK).  The Keep has an agent running which is listening on a
+well-known port, or on a port number previously established during
+the Keep preparation phase.  The ECA connects to this port, and
+establishes a connection to the Keep.  The Keep takes the workload,
+decrypts it, and runs it.
```

Comment:
May there be use-cases where it decrypts it but doesn't run it immediately? E.g., might the tenant want to measure the Keep after loading the workload but before running it?

Reply:
I don't this is excluded by the text as it stands.  However, I don't think your particular use case is likely: the tenant should not send the workload to the Keep until it is confident that it has been successfully attested.

Reply:
I'm personally interested to have the Keep state verified after the workload is loaded. But I think it doesn't have to be part of this RFC. How about removing anything from this RFC after the workload has been received and decrypted by the Keep?

Reply:
Do you mean the tenant workload?  That's what this RFC is referring to.  If so, that's difficult: though we might be able to measure the Keep+workload, we are unlikely to have a reference measurement against which to check it.